RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc

```diff
@@ -0,0 +1,419 @@
+    throw cms::Exception("MVA config failure: ")
+      << "wrong number of weightfiles" << std::endl;
+  }
+
+  gbrForests_.clear();
+  // Create a TMVA reader object for each category
+  for(int i=0; i<nCategories_; i++){
+
+    // Use unique_ptr so that all readers are properly cleaned up
+    // when the vector clear() is called in the destructor
+
+    edm::FileInPath weightFile( weightFileNames[i] );
+    gbrForests_.push_back( GBRForestTools::createGBRForest( weightFile ) );
+  }
+
+  withIso_ = withIso;
+}
+
+ElectronMVAEstimatorRun2Fall17::
+~ElectronMVAEstimatorRun2Fall17(){
+}
+
+void ElectronMVAEstimatorRun2Fall17::setConsumes(edm::ConsumesCollector&& cc) const {
+
+  // All tokens for event content needed by this MVA
+
+  // Beam spot (same for AOD and miniAOD)
+  //beamSpotToken_           = cc.consumes<reco::BeamSpot>(beamSpotLabel_);
+  cc.consumes<reco::BeamSpot>(beamSpotLabel_);
+
+  // Conversions collection (different names in AOD and miniAOD)
+  //conversionsTokenAOD_     = cc.mayConsume<reco::ConversionCollection>(conversionsLabelAOD_);
+  //conversionsTokenMiniAOD_ = cc.mayConsume<reco::ConversionCollection>(conversionsLabelMiniAOD_);
+  cc.mayConsume<reco::ConversionCollection>(conversionsLabelAOD_);
+  cc.mayConsume<reco::ConversionCollection>(conversionsLabelMiniAOD_);
+
+  //rhoToken_                = cc.consumes<double>(rhoLabel_);
+  cc.consumes<double>(rhoLabel_);
+
+}
+
+float ElectronMVAEstimatorRun2Fall17::
+mvaValue( const edm::Ptr<reco::Candidate>& particle, const edm::Event& iEvent) const {
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
+  if( eleRecoPtr.get() == nullptr ) {
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+  }
+
+  const int iCategory = findCategory( eleRecoPtr.get() );
+  const std::vector<float> vars = fillMVAVariables( particle, iEvent );
+  return mvaValue(iCategory, vars);
+}
+
+float ElectronMVAEstimatorRun2Fall17::
+mvaValue( const reco::GsfElectron * particle, const edm::EventBase & iEvent) const {
+  edm::Handle<reco::ConversionCollection> conversions;
+  edm::Handle<reco::BeamSpot> beamSpot;
+  edm::Handle<double> rho;
+  iEvent.getByLabel(conversionsLabelAOD_, conversions);
+  iEvent.getByLabel(beamSpotLabel_, beamSpot);
+  iEvent.getByLabel(rhoLabel_, rho);
+  const int iCategory = findCategory( particle );
+  const std::vector<float> vars = fillMVAVariables( particle, conversions, beamSpot.product(), rho );
+  return mvaValue(iCategory, vars);
+}
+
+float ElectronMVAEstimatorRun2Fall17::
+mvaValue( const int iCategory, const std::vector<float> & vars) const  {
+  const float result = gbrForests_.at(iCategory)->GetClassifier(vars.data());
+
+  const bool debug = false;
```

Comment:
use a preprocessor directive to enclose the following debug lines, or at least make ```debug``` a constexpr

Reply:
I thought preprocessor directives are not welcome in cmssw? How would you do it? I made debug a constexpr for now.